### PR TITLE
doc: Re-add C++ linkage back to gendocfile

### PR DESCRIPTION
**Pull request by ibuclaw**
As this is called from the compiler main function.

### CODE REVIEW

**Review by Geod24 — CHANGES_REQUESTED** [review]
If it's called from C++, can you add it to a header file?

**Comment by ibuclaw** [comment]
For the docs module, this would be `gendocfile` at the top and public, the rest of the functions and declarations that follow are `package`, everything else that is already `private` gets grouped together in the appropriate place - probably the bottom.

**Comment by jacob-carlborg** [comment]
> In this case, gendocfile is only used by mars.d - which is NOT a front end module. :-)

So why does it need to be `extern(C++)`? I thought `mars.d` was only used by DMD.

**Comment by ibuclaw** [comment]
@jacob-carlborg - Because dmd is not the only compiler using the front end...

**Comment by jacob-carlborg** [comment]
@ibuclaw you said `mars.d` is not a front end module :wink:.

**Comment by Geod24** [comment]
@ibuclaw : If it is `extern(C++)`, it's because it has to be called from C++... If it has to be called from C++, it should be in the header. Yes they are not accurate at the moment, but there's no reason to keep it that way.

**Comment by ibuclaw** [comment]
@Geod24 - created a lonely file.

**Comment by ibuclaw** [comment]
> you said mars.d is not a front end module wink.

That is correct, it is not a front end module.  The best label to give it would be the driver, parts of the compiler that deal with inputs and outputs.  Other modules with the same label include dinifile, libXXX, link, and scanXXX.

File: src/dmd/doc.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * Distributed under the Boost Software License, Version 1.0.
+ * http://www.boost.org/LICENSE_1_0.txt
+ * https://github.com/dlang/dmd/blob/master/src/dmd/doc.h
+ */
+
+#ifndef DMD_DOC_H
+#define DMD_DOC_H
+
+#ifdef __DMC__
+#pragma once
```
Comment (Geod24) [inline src/dmd/doc.h]: Isn't pragma once implemented by all compilers? :D

**Review by Geod24 — APPROVED** [review]
LGTM

File: src/dmd/doc.h (hunk 1 above)
Comment (ibuclaw) [reply src/dmd/doc.h]: Let's continue this in #8641

**Pull request merged**